### PR TITLE
[JENKINS-46175] SimpleBuildStep/SimpleBuildWrapper without workspace

### PR DESCRIPTION
**Pull request by Zastai**
See [JENKINS-46175](https://issues.jenkins-ci.org/browse/JENKINS-46175).

This makes the workspace and launcher parameters nullable for
`SimpleBuildStep.perform()` (both overloads).

New `requiresLauncher()` and `requiresWorkspace()` predicates determine
whether they are really required; by default, these return `true`,
preserving the old non-null behaviour.

Similar changes were made for `SimpleBuildWrapper`, with the predicates
controlling the nullability of workspace/launcher on the `setUp` method.

Currently no tests added; this does not really change any behaviour, it only
enables plugins not to require a workspace/launcher context. It is `CoreStep`
and `CoreWrapperStep` in `workspace-basic-steps` that will get behaviour
changes. Setting this as draft until I have a PR ready on that plugin (using
an incremental from this one).

Note: initially I just had a single `requiresWorkspace()` for both; but then I saw
that `BuildStepCompatibilityLayer` seemed to support a case where workspace
and launcher were not present in lockstep, I added a separate predicate. This
seems harmless and allows more granular specification of needs.

### Proposed changelog entries

* JENKINS-46175: A `SimpleBuildStep` or `SimpleBuildWrapper` can now choose not to require a workspace and/or launcher.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

Reference implementation:

* https://github.com/jenkinsci/workflow-basic-steps-plugin/pull/117

### Desired reviewers

@mention jglick

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by Zastai** [comment]
The behaviour is tested as part of https://github.com/jenkinsci/workflow-basic-steps-plugin/pull/117, where it's `CoreStep` that will allow running `SimpleBuildStep`s outside of node context (and same for `CoreWrapper`).
See https://github.com/Zastai/workflow-basic-steps-plugin/blob/JENKINS-46175/src/test/java/org/jenkinsci/plugins/workflow/steps/CoreStepTest.java#L172-L220 and https://github.com/Zastai/workflow-basic-steps-plugin/blob/JENKINS-46175/src/test/java/org/jenkinsci/plugins/workflow/steps/CoreWrapperStepTest.java#L293-L363

I'm not entirely sure I could test this in Jenkins core; I'd have to see if there's anything in the test harness that allows me to run something other than `AbstractProject`, in order to call a step without passing workspace/launcher. But then I'm not familiar with the entire codebase yet, so maybe there is a way.

File: core/src/main/java/hudson/tasks/BuildStepCompatibilityLayer.java
Comment (jglick) [inline core/src/main/java/hudson/tasks/BuildStepCompatibilityLayer.java:76]: Note that this would be a pathological condition: the build is almost certainly broken anyway.

File: core/src/main/java/hudson/tasks/BuildStepCompatibilityLayer.java
Comment (jglick) [inline core/src/main/java/hudson/tasks/BuildStepCompatibilityLayer.java:80]: Similarly, for an `AbstractBuild` this would be a pathological condition. There is probably no real need to patch this class at all.

File: core/src/main/java/jenkins/tasks/SimpleBuildStep.java
Comment (jglick) [inline core/src/main/java/jenkins/tasks/SimpleBuildStep.java:89]:
```suggestion
    default void perform(@NonNull Run<?, ?> run, @Nullable FilePath workspace, @Nullable Launcher launcher,
```

since whether or not null is actually allowed is dependent on complex things the static analyzer will not understand

File: core/src/main/java/jenkins/tasks/SimpleBuildStep.java
Comment (jglick) [inline core/src/main/java/jenkins/tasks/SimpleBuildStep.java:106]:
```suggestion
    default void perform(@NonNull Run<?, ?> run, @Nullable FilePath workspace, @NonNull EnvVars env, @Nullable Launcher launcher,
```

File: core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java
Comment (jglick) [inline core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:78]:
```suggestion
    public abstract void setUp(@NonNull Context context, @NonNull Run<?,?> build, @Nullable FilePath workspace, @Nullable Launcher launcher, @NonNull TaskListener listener, @NonNull EnvVars initialEnvironment) throws IOException, InterruptedException;
```

File: core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java
Comment (jglick) [inline core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:155]:
```suggestion
        public abstract void tearDown(@NonNull Run<?,?> build, @Nullable FilePath workspace, @Nullable Launcher launcher, @NonNull TaskListener listener) throws IOException, InterruptedException;
```

File: core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java
Comment (jglick) [inline core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:181]: Redundant. Either the wrapper setup & teardown both require a node context (~ workspace + launcher), or neither do.

File: core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java
Comment (jglick) [inline core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:193]:
```suggestion
    @Override public final Environment setUp(@NonNull AbstractBuild build, final @Nullable Launcher launcher, @NonNull BuildListener listener) throws IOException, InterruptedException {
```

File: core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java
Comment (jglick) [inline core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:203]:
```suggestion
    @Override public final void preCheckout(@NonNull AbstractBuild build, final @Nullable Launcher launcher, @NonNull BuildListener listener) throws IOException, InterruptedException {
```

File: core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java
Comment (jglick) [inline core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:213]:
```suggestion
        private final @Nullable Launcher launcher;
```

File: core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java
Comment (jglick) [inline core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:214]:
```suggestion
        EnvironmentWrapper(@NonNull Context c, @Nullable Launcher launcher) {
```

File: core/src/main/java/jenkins/tasks/SimpleBuildStep.java
Comment (Zastai) [reply core/src/main/java/jenkins/tasks/SimpleBuildStep.java:89]: It was not particularly clear to me when each of those two was applicable; I rather assumed they were equivalent so I just used the same one everywhere. Will have a look around and adjust these.

Comment (jglick) [reply core/src/main/java/jenkins/tasks/SimpleBuildStep.java:89]: > The annotated element could be null under some circumstances.
> In general, this means developers will have to read the documentation to determine when a null value is acceptable and whether it is necessary to check for a null value.

**Comment by jglick** [comment]
https://github.com/jenkinsci/jenkins/pull/4784#discussion_r445877019 should be addressed, I think: the API is unnecessarily broad. Everything else looks fine.

File: core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java
Comment (Zastai) [reply core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:181]: The `Disposer` class is public API, not restricted to `SimpleBuildWrapper`s; so I figured it made sense to give it support for separate configuration.

Also, without these methods, there is nothing to refer to in the tearDown() JavaDoc to explain when the argument may be null, because there is no link to the original wrapper in this class.

An org search suggests there is no plugin using `Disposer` outside a wrapper context, so if you're happy with just making the workspace/launcher nullable, then I can indeed drop these.

File: core/src/main/java/hudson/tasks/BuildStepCompatibilityLayer.java
Comment (Zastai) [reply core/src/main/java/hudson/tasks/BuildStepCompatibilityLayer.java]: File changes reverted.

File: core/src/main/java/jenkins/tasks/SimpleBuildStep.java
Comment (Zastai) [reply core/src/main/java/jenkins/tasks/SimpleBuildStep.java]: Switched `@CheckForNull` to `@Nullable` as suggested

File: core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java
Comment (jglick) [reply core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:182]: Yes a disposer matches a wrapper. It is only a separate type for reasons of serialization: you must carefully manage state of the disposer.

Comment (Zastai) [reply core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java]: Methods dropped; JavaDoc adjusted.

File: core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java
Comment (jglick) [inline core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:151]:
```suggestion
         * @param workspace a workspace of the build; may be {@code null} if allowed by the corresponding wrapper
         * @param launcher a way to start commands; may be {@code null} if allowed by the corresponding wrapper
```

**Review by jglick — APPROVED** [review]

File: core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java
Comment (Zastai) [reply core/src/main/java/jenkins/tasks/SimpleBuildWrapper.java:151]: Yes, I seem to have either lost an `it` or failed to remove the `is`.

**Review by oleg-nenashev — COMMENTED** [review]
I am not a big fan of using `Nullable` in API, because it does not really enforce any checks in the static analysis tools. TBH My preference would be to introduce new API endpoints or new child classes for workspace-less operations, but it could also overcomplicate APi as well. I am not bloking this pull request though

**Comment by Zastai** [comment]
I'll look at drafting an alternative PR that adds extra API instead.

**Comment by jglick** [comment]
`@Nullable` is not ideal, but on balance I think this is preferable to #4829. No strong opinion.

**Review by oleg-nenashev — APPROVED** [review]
I prefer #4829, but this one will do the job as well. I approve this pull request to indicate that I can live with that if other core maintainers prefer this one

**Review by MRamonLeon — COMMENTED** [review]
I would like to see some checks for the `Nullable` parameters. If they're null, then the `requiresXXX` method has to return true. Other way, it's up to the developer. And we don't trust developers :wink: 

I like more #4829 for this reason.

**Comment by Zastai** [comment]
Closing - this has been replaced by https://github.com/jenkinsci/jenkins/pull/4829

File: core/src/main/java/jenkins/tasks/SimpleBuildStep.java
Comment (reinholdfuereder) [inline core/src/main/java/jenkins/tasks/SimpleBuildStep.java:139]: It looks like there still 3 little open TODOs in this PR, cf. "@since TODO"

Comment (Zastai) [reply core/src/main/java/jenkins/tasks/SimpleBuildStep.java:139]: Aside from this PR being closed, those are to be the version number of the first release that includes the API. Because this is a moving target, it is usually left to be filled in after the PR is merged.

**Pull request closed without merging**
